main/src/ca/uwaterloo/flix/api/Flix.scala

```diff
@@ -485,84 +487,126 @@ class Flix {
   /**
     * Compiles the Flix program and returns a typed ast.
     */
-  def check(): Validation[TypedAst.Root, CompilationMessage] = try {
-    import Validation.Implicit.AsMonad
+  def check(): (Validation[TypedAst.Root, CompilationMessage], List[CompilationMessage & Recoverable]) = {
```

Comment:
I think when making big changes like this it is better to discuss first.

Reply:
I can imagine two better signatures:

- Validation[(TypedAst.Root, Message), Message]
- (Option[TypedAst.Root], List[Message])

Reply:
What are your thoughts on these?

Reply:
This is also just a proposal.

I see a couple of options:
a) This solution: return a Validation along with recoverable failures which the caller can decide what to do with. Only return Hard failure if the front-end produced one.
b) Only return success if no error was present at all (then we should probably add a separate test function for `TestParser`)

Reply:
I don't understand why it is just TestParser- should all the frontend test not just call check?

Reply:
And what about the convenience functions we have, e.g. expectError etc?

Reply:
I have to, but could you write up a small blob of text to outline the problem and discuss the trade-offs of the different signatures?

Reply:
Most tests just call `compile` because will catch the expected error anyway. The way I understand it is that the intention was that now that the parser is resilient, we should check that the front-end can succeed. Of course we should probably always check that but since most tests only really care about producing the expected error it does not really make a difference.

Reply:
OK, but what are the implications of that?